lib/system.nim

```diff
@@ -3007,3 +3006,25 @@ export io
 
 when not defined(createNimHcr):
   include nimhcr
+
+when defined(nimHasInternalProc):
+  ## {.internalproc.} is analog to {.compilerproc.} but is used for semantic
+  ## phase instead of codegen. It simplifies implementation by using nim code
+  ## instead of building the AST by hand, while also avoiding making these
+  ## public.
+  proc nimInternalNewException(name: string, msg: string, traceMsg: string) {.internalproc.} =
+    ## internal proc for raising from vmpops
+    template fun(T) =
+      if name == astToStr(T):
+        var msg = msg
+        if traceMsg.len > 0:
+          # we could query for whether nim was build with --stacktrace to
+          # provide more accurate msg, but not sure if it's exposed.
+          msg.add "\nVM callback stacktrace (compile nim with --stacktrace if needed)\n" & traceMsg & "\n"
+        raise newException(T, msg)
+    # can add more hardcoded standard types here
+    fun(OSError)
+    fun(IOError)
+    fun(AssertionError)
```

Comment:
Does thins mean that only `OSError`, `IOError` and `AssertionError` can be raised? Isn't there space to have a generic solution here that works on all error kinds?

Reply:
If it is not possible to use compilerProc here then it needs to be moved into compiler